mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageIdMapperTest.java

```diff
@@ -0,0 +1,413 @@
+            .add(Flag.RECENT)
+            .add(Flag.ANSWERED)
+            .build();
+        int modSeq = 2;
+        UpdatedFlags expectedUpdatedFlags = new UpdatedFlags(message1.getUid(), modSeq, initialFlags, newFlags);
+        assertThat(flags).containsOnly(MapEntry.entry(benwaInboxMailbox.getMailboxId(), expectedUpdatedFlags));
+    }
+
+    @ContractTest
+    public void setFlagsShouldReturnUpdatedFlagsWhenMessageIsInTwoMailboxes() throws Exception {
+        message1.setUid(mapperProvider.generateMessageUid());
+        sut.save(message1);
+
+        SimpleMailboxMessage message1InOtherMailbox = SimpleMailboxMessage.copy(benwaWorkMailbox.getMailboxId(), message1);
+        message1InOtherMailbox.setUid(mapperProvider.generateMessageUid());
+        sut.save(message1InOtherMailbox);
+
+        MessageId messageId = message1.getMessageId();
+        Flags newFlags = new Flags(Flag.ANSWERED);
+        Map<MailboxId, UpdatedFlags> flags = sut.setFlags(newFlags, MessageManager.FlagsUpdateMode.REMOVE, messageId);
+
+        int modSeq = 1;
+        UpdatedFlags expectedUpdatedFlags = new UpdatedFlags(message1.getUid(), modSeq, new Flags(), newFlags);
+        UpdatedFlags expectedUpdatedFlags2 = new UpdatedFlags(message1InOtherMailbox.getUid(), modSeq, new Flags(), newFlags);
+        assertThat(flags).containsOnly(MapEntry.entry(benwaInboxMailbox.getMailboxId(), expectedUpdatedFlags),
+                MapEntry.entry(benwaWorkMailbox.getMailboxId(), expectedUpdatedFlags2));
+    }
+
+    @ContractTest
+    public void setFlagsShouldUpdateFlagsWhenMessageIsInOneMailbox() throws Exception {
+        message1.setUid(mapperProvider.generateMessageUid());
+        sut.save(message1);
+
+        MessageId messageId = message1.getMessageId();
+        sut.setFlags(new Flags(Flag.ANSWERED), MessageManager.FlagsUpdateMode.REMOVE, messageId);
+
+        List<MailboxMessage> messages = sut.find(ImmutableList.of(messageId), MessageMapper.FetchType.Body);
+        assertThat(messages).hasSize(1);
+        assertThat(messages.get(0).isAnswered()).isTrue();
+    }
+
+    @ContractTest
+    public void setFlagsShouldUpdateFlagsWhenMessageIsInTwoMailboxes() throws Exception {
+        message1.setUid(mapperProvider.generateMessageUid());
+        sut.save(message1);
+
+        SimpleMailboxMessage message1InOtherMailbox = SimpleMailboxMessage.copy(benwaWorkMailbox.getMailboxId(), message1);
+        message1InOtherMailbox.setUid(mapperProvider.generateMessageUid());
+        sut.save(message1InOtherMailbox);
+
+        MessageId messageId = message1.getMessageId();
+        sut.setFlags(new Flags(Flag.ANSWERED), MessageManager.FlagsUpdateMode.REMOVE, messageId);
+
+        List<MailboxMessage> messages = sut.find(ImmutableList.of(messageId), MessageMapper.FetchType.Body);
+        assertThat(messages).hasSize(2);
+        assertThat(messages.get(0).isAnswered()).isTrue();
+        assertThat(messages.get(1).isAnswered()).isTrue();
+    }
+
```

Comment:
Verify also modseq increment upon setFlags (It's mapper reponsibity to increment it on each retry...)